### ISSUE TITLE
$jacocoInit() class cast exception

### ISSUE THREAD

**Issue by xyhuangjinfu** [jacoco/jacoco#848]
This is a issue tracker. Please use our mailing list for general questions: 
https://groups.google.com/forum/?fromgroups=#!forum/jacoco

Also check FAQ before opening an issue: http://www.jacoco.org/jacoco/trunk/doc/faq.html

### Steps to reproduce

JaCoCo version: 0.8.3
Operating system: mac os 10.13.4
Tool integration: Maven/Ant/API/Other

When I instrumentd a class which with a public static main method, I saved this class into disk, and later I want to run this class via java command, I get the ClassCastException.
Exception info
`Exception in thread "main" java.lang.ClassCastException: java.lang.Long cannot be cast to [Z
	at Test.$jacocoInit(Test.java)
	at Test.main(Test.java)` 

### Expected behaviour
I want it can run normally, or if consider i lack some dependencies jar, it may throw ClassNotFoundException. why can i get a ClassCastException?

### Actual behaviour
I get a ClassCastException.

**Comment by Godin** [comment]
How exactly you do all this? Please provide complete example.

**Comment by xyhuangjinfu** [comment]
Thanks for response, now I'll give a full reproduce process. steps are below:
_note: I'm work on windows os now, not mac os, but this not effect the result. jacoco version is still 0.8.3._
1. I have a demo class which print hello word.
```
public class Test {
    public static void main(String[] args) {
        System.out.print("hello world");
    }
}
```
2. compile it with javac command and run it with java command, and it print "hello world"
```
PS C:\Users\huangjinfu\Desktop\xx> javac Test.java
PS C:\Users\huangjinfu\Desktop\xx> java Test
hello world
```
3. I use jacoco to instrument it, the code to instrument this class are from jacoco CoreTutorial.
[https://www.jacoco.org/jacoco/trunk/doc/examples/java/CoreTutorial.java](https://www.jacoco.org/jacoco/trunk/doc/examples/java/CoreTutorial.java)
```
    public static void main(String[] args) {
        try {
            String originClassFilePath = "C:\\Users\\huangjinfu\\Desktop\\xx\\Test.class";

            IRuntime runtime = new LoggerRuntime();
            Instrumenter instrumenter = new Instrumenter(runtime);

            FileInputStream fis = new FileInputStream(originClassFilePath);
            byte[] instrumented = instrumenter.instrument(fis, "error msg");
            fis.close();

            FileOutputStream fos = new FileOutputStream(originClassFilePath);
            fos.write(instrumented);
            fos.flush();
            fos.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
```
4. now I execute java command to run this class, it throw a ClassCastException.
```
PS C:\Users\huangjinfu\Desktop\xx> java Test
二月 23, 2019 9:18:40 上午 Test $jacocoInit
信息: 4dc63996
Exception in thread "main" java.lang.ClassCastException: java.lang.Long cannot be cast to [Z
        at Test.$jacocoInit(Test.java)
        at Test.main(Test.java)
```
5. Class info before and after instrumented are below
```
//before instrumented
PS C:\Users\huangjinfu\Desktop\xx> javap -c Test
Compiled from "Test.java"
public class Test {
  public Test();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  public static void main(java.lang.String[]);
    Code:
       0: getstatic     #2                  // Field java/lang/System.out:Ljava/io/PrintStream;
       3: ldc           #3                  // String hello world
       5: invokevirtual #4                  // Method java/io/PrintStream.print:(Ljava/lang/String;)V
       8: return
}

//after instrumented
PS C:\Users\huangjinfu\Desktop\xx> javap -c Test
Compiled from "Test.java"
public class Test {
  public Test();
    Code:
       0: invokestatic  #32                 // Method $jacocoInit:()[Z
       3: astore_1
       4: aload_0
       5: invokespecial #1                  // Method java/lang/Object."<init>":()V
       8: aload_1
       9: iconst_0
      10: iconst_1
      11: bastore
      12: return

  public static void main(java.lang.String[]);
    Code:
       0: invokestatic  #32                 // Method $jacocoInit:()[Z
       3: astore_1
       4: getstatic     #2                  // Field java/lang/System.out:Ljava/io/PrintStream;
       7: ldc           #3                  // String hello world
       9: invokevirtual #4                  // Method java/io/PrintStream.print:(Ljava/lang/String;)V
      12: aload_1
      13: iconst_1
      14: iconst_1
      15: bastore
      16: return
}
```

**Comment by Godin** [comment]
Instrumented classes can't be used alone, they require exact same `IRuntime` which was used for their instrumentation (in this case `LoggerRuntime`), and this `IRuntime` must be started as shown in `CoreTutorial`:

https://github.com/jacoco/jacoco/blob/v0.8.3/org.jacoco.examples/src/org/jacoco/examples/CoreTutorial.java#L120-L123

And this is the case for all implementations of `IRuntime`.

> I use jd-gui to see the instrumented class file, not find something related to Long type

Because both `javap -c` and `jd-gui` do not show you everything. In particular implementation of `$jacocoInit` method added by instrumentation. Use `javap` with `-p` (Show all classes and members):

```
  private static boolean[] $jacocoInit();
    descriptor: ()[Z
    flags: ACC_PRIVATE, ACC_STATIC, ACC_SYNTHETIC
    Code:
      stack=5, locals=0, args_size=0
         0: getstatic     #37                 // Field $jacocoData:[Z
         3: dup
         4: ifnonnull     60
         7: pop
         8: iconst_3
         9: anewarray     #4                  // class java/lang/Object
        12: dup
        13: iconst_0
        14: ldc2_w        #38                 // long 3367819320579584995l
        17: invokestatic  #45                 // Method java/lang/Long.valueOf:(J)Ljava/lang/Long;
        20: aastore
        21: dup
        22: iconst_1
        23: ldc           #46                 // String org/jacoco/examples/CoreTutorial$TestTarget
        25: aastore
        26: dup
        27: iconst_2
        28: bipush        6
        30: invokestatic  #51                 // Method java/lang/Integer.valueOf:(I)Ljava/lang/Integer;
        33: aastore
        34: dup
        35: ldc           #53                 // String jacoco-runtime
        37: invokestatic  #59                 // Method java/util/logging/Logger.getLogger:(Ljava/lang/String;)Ljava/util/logging/Logger;
        40: swap
        41: getstatic     #65                 // Field java/util/logging/Level.INFO:Ljava/util/logging/Level;
        44: swap
        45: ldc           #67                 // String 4dc63996
        47: swap
        48: invokevirtual #71                 // Method java/util/logging/Logger.log:(Ljava/util/logging/Level;Ljava/lang/String;[Ljava/lang/Object;)V
        51: iconst_0
        52: aaload
        53: checkcast     #72                 // class "[Z"
        56: dup
        57: putstatic     #37                 // Field $jacocoData:[Z
        60: areturn
```

> Why i get a ClassCastException?

From `$jacocoInit` provided above should be already clear. However should be noted: when using something, please at least read its documentation, or even better - implementation. If you'll read documentation of `LoggerRuntime`, then, I hope, you'll understand why classes instrumented with `LoggerRuntime` throw `ClassCastException` when `LoggerRuntime` not started. Other implementations of `IRuntime` may throw other exceptions in such case.

You can also read https://www.jacoco.org/jacoco/trunk/doc/implementation.html (in particular "Coverage Runtime Dependency") that states:

> Making a runtime library available to all instrumented classes can be a painful or impossible task in frameworks that use their own class loading mechanisms.

In other words - `ClassLoader`s can restrict direct access of instrumented classes to class of `IRuntime` and that's why such access is avoided. No direct access = no automatic throwing of `ClassNotFoundException`. Moreover `ClassNotFoundException` anyway does not make sense in case when start of `IRuntime` forgotten. Addition of detection of such case increases complexity and will increase size of instrumented classes and so ROI of this IMO is highly questionable.

And IMO usage of low level abstractions always requires understanding of their requirements. Or just use high level abstracts such as JaCoCo Java Agent, which guarantees that runtime is started and instrumented classes can access it.
